[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/hien078/applied-mathematics-foundation/blob/master/probability_statistics/08_law_of_large_numbers_and_clt/exercises.ipynb)

# Exercises — Module 08: Law of Large Numbers and the Central Limit Theorem

20 fully solved problems in 4 levels: Concept Check (4), Foundation (6), Applications in AI/ML & Physics (6), Challenge (4).

In [1]:
import numpy as np
import matplotlib.pyplot as plt
from scipy.stats import norm

plt.rcParams.update({
    "figure.figsize": (7.0, 4.0),
    "figure.dpi": 110,
    "axes.grid": True,
    "grid.alpha": 0.3,
    "axes.spines.top": False,
    "axes.spines.right": False,
    "font.size": 11,
})
rng = np.random.default_rng(0)

## L0 — Concept Checks

### Problem L0.1 — Weak versus Strong — What Actually Differs?

State the weak and strong laws of large numbers precisely, and describe a hypothetical sequence of averages that would satisfy the weak law but violate the strong law. Why is no such sequence produced by i.i.d. integrable summands?

**Intuition.** The weak law controls each snapshot on its own; the strong law controls the whole trajectory at once. The gap between them is the gap between a sequence of probabilities tending to zero and the same sequence being summable.

**Solution**

**Weak LLN.** For every $\varepsilon \gt 0$,

$$
\lim_{n \to \infty} P\left(\left\vert \bar{X}_n - \mu \right\vert \gt \varepsilon\right) = 0 .
$$

This constrains each *individual* probability in the sequence; it says nothing about whether excursions recur.

**Strong LLN.** The set of sample paths that converge has full measure,

$$
P\left(\left\{\omega : \lim_{n\to\infty}\bar{X}_n(\omega) = \mu\right\}\right) = 1 ,
$$

equivalently, for every $\varepsilon \gt 0$ the event $\{\vert \bar{X}_n - \mu\vert \gt \varepsilon\}$ occurs only finitely often, almost surely.

**A separating example.** Let $Y_n$ be *independent* with $P(Y_n = 1) = 1/n$ and $P(Y_n = 0) = 1 - 1/n$. Then $P(\vert Y_n - 0\vert \gt \varepsilon) = 1/n \to 0$, so $Y_n \xrightarrow{P} 0$. But $\sum_n P(Y_n = 1) = \sum_n 1/n = \infty$ with independence, so the second Borel–Cantelli lemma gives $P(Y_n = 1 \text{ i.o.}) = 1$: the sequence spikes to $1$ infinitely often and does **not** converge almost surely. Convergence in probability without almost sure convergence is exactly the failure of summability.

**Why averages of i.i.d. data never behave this way.** Kolmogorov's theorem shows that for i.i.d. summands with $E\vert X_1\vert \lt \infty$, the excursion probabilities decay fast enough (after maximal-inequality blocking) to be summable along dyadic blocks, forcing almost sure convergence. Integrability is the sharp condition: if $E\vert X_1\vert = \infty$, then $\limsup_n \vert \bar{X}_n\vert = \infty$ a.s.

$$
\boxed{\text{Weak} = \text{each } P(\text{bad}_n) \to 0; \quad \text{Strong} = P(\text{bad}_n \text{ i.o.}) = 0}
$$

*Key takeaway:* The gap between the two laws is the gap between a sequence tending to zero and a series converging.

### Problem L0.2 — What the CLT Does and Does Not Say

A student writes: "By the CLT, for large $n$ the sample mean $\bar{X}_n$ converges to a normal distribution." Identify the two errors and give the correct statement, including why the factor $\sqrt{n}$ is the unique correct scaling.

**Intuition.** The sample mean collapses onto a single point, so anything with a nondegenerate limit must be the deviation magnified by exactly the rate at which that deviation shrinks.

**Solution**

**Error 1 — the object.** $\bar{X}_n$ itself converges to the *constant* $\mu$ (LLN), so its limiting law is a point mass, not a normal law. The CLT concerns the centered and rescaled deviation.

**Error 2 — the mode.** Even for the rescaled quantity, convergence is *in distribution*, not pointwise or almost sure. Nothing is claimed about the random variables converging, only about their CDFs.

**Correct statement.**

$$
Z_n = \frac{\sqrt{n}\left(\bar{X}_n - \mu\right)}{\sigma} \xrightarrow{d} \mathcal{N}(0,1), \qquad \text{equivalently} \qquad \bar{X}_n \overset{\text{approx}}{\sim} \mathcal{N}\!\left(\mu, \frac{\sigma^2}{n}\right) .
$$

**Uniqueness of the scaling.** Consider $n^{\alpha}(\bar{X}_n - \mu)$, whose variance is $n^{2\alpha}\sigma^2/n = \sigma^2 n^{2\alpha - 1}$. This limit is:

- $0$ if $\alpha \lt 1/2$ (over-shrunk: degenerate at $0$),
- $\infty$ if $\alpha \gt 1/2$ (over-magnified: no proper limit),
- $\sigma^2$ exactly when $\alpha = 1/2$.

Only $\alpha = 1/2$ balances the shrinking fluctuation against the magnification, which is why $\sqrt{n}$ is *the* rate.

$$
\boxed{\sqrt{n}\left(\bar{X}_n - \mu\right) \xrightarrow{d} \mathcal{N}(0, \sigma^2), \text{ while } \bar{X}_n \xrightarrow{P} \mu}
$$

*Key takeaway:* LLN kills the fluctuation; the CLT resurrects it under a microscope of magnification $\sqrt{n}$.

### Problem L0.3 — A Chebyshev Bound for Coin Flips

For $n = 100$ fair coin flips let $\hat{p}$ be the observed fraction of heads. Bound $P(\vert \hat{p} - 0.5\vert \ge 0.1)$ using Chebyshev, then compare with the CLT approximation. Comment on the gap.

**Intuition.** Chebyshev has to survive the worst law with the given variance, while the CLT already knows the shape is nearly Gaussian, so expect a factor of several between the two answers.

**Solution**

**Chebyshev.** Each flip is Bernoulli($0.5$) with $\sigma^2 = p(1-p) = 0.25$, so $\operatorname{Var}(\hat{p}) = 0.25/100 = 0.0025$ and

$$
P\left(\left\vert \hat{p} - 0.5 \right\vert \ge 0.1\right) \le \frac{0.0025}{0.1^2} = 0.25 .
$$

**CLT.** The standard error is $\operatorname{sd}(\hat p) = \sqrt{0.0025} = 0.05$, so the threshold sits at $0.1/0.05 = 2$ standard errors:

$$
P\left(\left\vert \hat{p} - 0.5\right\vert \ge 0.1\right) \approx 2\left(1 - \Phi(2)\right) = 2(0.02275) = 0.0455 .
$$

(The exact binomial answer, $P(\vert B - 50\vert \ge 10)$ with $B \sim \text{Bin}(100, 0.5)$, is $\approx 0.0569$; with a continuity correction at $2.5$ standard errors minus half a count the normal approximation lands near $0.057$.)

**The gap.** Chebyshev overestimates the exact answer by a factor $0.25/0.0569 \approx 4.4$ because it uses only two moments and must remain valid for the worst distribution with that variance — e.g. a three-point law placing all excess mass exactly at $\pm\varepsilon$. The CLT exploits the actual shape (here, near-Gaussian), obtaining exponential rather than $\varepsilon^{-2}$ decay.

$$
\boxed{\text{Chebyshev} \le 0.25, \quad \text{CLT} \approx 0.046, \quad \text{exact} \approx 0.057}
$$

*Key takeaway:* Chebyshev is a guarantee (valid always, loose usually); the CLT is an approximation (sharp usually, silent about validity).

Recompute the Chebyshev bound, the CLT approximation and the exact binomial probability.

In [2]:
from scipy.stats import binom

n_flips, eps0 = 100, 0.1
cheb = (0.25 / n_flips) / eps0**2
clt = 2 * (1 - norm.cdf(eps0 / np.sqrt(0.25 / n_flips)))
exact = 2 * float(binom.cdf(40, n_flips, 0.5))      # |B - 50| >= 10
print(f"Chebyshev = {cheb:.4f}   CLT = {clt:.4f}   exact = {exact:.4f}")
print(f"Chebyshev / exact = {cheb / exact:.2f}   CLT / exact = {clt / exact:.2f}")
assert abs(cheb - 0.25) < 1e-12
assert abs(clt - 0.0455) < 5e-4
assert abs(exact - 0.0569) < 5e-4

Chebyshev = 0.2500   CLT = 0.0455   exact = 0.0569
Chebyshev / exact = 4.39   CLT / exact = 0.80


The three numbers of the boxed answer are confirmed. Chebyshev is $4.4$ times the truth
because it must hold for every law with variance $0.25$; the CLT approximation is $20\%$ low
because it ignores the lattice structure that a continuity correction would restore.

### Problem L0.4 — Which Distributions Are Excluded?

For each summand law, state whether the weak LLN, strong LLN, and CLT apply: (a) $\text{Bernoulli}(0.3)$; (b) $\text{Exponential}(\lambda)$; (c) Pareto with density $f(x) = 1.5 x^{-2.5}$ on $x \gt 1$; (d) standard Cauchy.

**Intuition.** Two integrals decide everything here: $E\vert X\vert$ governs both laws of large numbers, and $E[X^2]$ governs the CLT.

**Solution**

Check the two thresholds: $E\vert X\vert \lt \infty$ (both LLNs) and $E[X^2] \lt \infty$ (CLT).

**(a) Bernoulli($0.3$).** Bounded, so all moments are finite: WLLN ✓, SLLN ✓, CLT ✓ with $\mu = 0.3$, $\sigma^2 = 0.21$.

**(b) Exponential($\lambda$).** All moments finite ($E[X^k] = k!/\lambda^k$): WLLN ✓, SLLN ✓, CLT ✓ with $\mu = 1/\lambda$, $\sigma^2 = 1/\lambda^2$.

**(c) Pareto with tail index $\alpha = 1.5$.** Here $f(x) = \alpha x^{-(\alpha+1)}$ with $\alpha = 1.5$. Then

$$
E[X] = \int_1^\infty x \cdot 1.5x^{-2.5}\,dx = 1.5\int_1^\infty x^{-1.5}\,dx = 3 \lt \infty ,
$$

but

$$
E[X^2] = 1.5\int_1^\infty x^{-0.5}\,dx = \infty .
$$

So WLLN ✓, SLLN ✓, **CLT ✗** — the correct limit is a $1.5$-stable law with scaling $n^{1/1.5} = n^{2/3}$, not $\sqrt{n}$.

**(d) Standard Cauchy.** $E\vert X\vert = \frac{2}{\pi}\int_0^\infty \frac{x}{1+x^2}dx = \infty$: **all three fail**. Indeed $\bar{X}_n$ is exactly standard Cauchy for every $n$.

$$
\boxed{\text{(a),(b): all three; (c): LLNs only; (d): none}}
$$

*Key takeaway:* One integral of $\vert x\vert$ and one of $x^2$ decide which limit theorems are available.

## L1 — Foundations

### Problem L1.1 — Sample Size from Chebyshev, Hoeffding, and the CLT

A pollster wants $P(\vert \hat{p} - p\vert \ge 0.03) \le 0.05$ for an unknown proportion $p$. Determine the required $n$ from (a) Chebyshev, (b) Hoeffding, (c) the CLT, using the worst case over $p$.

**Intuition.** All three answers scale as $\varepsilon^{-2}$; what separates them is how the failure probability enters — as $1/\delta$, as $\ln(1/\delta)$, or through a Gaussian quantile.

**Solution**

The worst-case variance for a Bernoulli sample is $p(1-p) \le 1/4$, attained at $p = 1/2$.

**(a) Chebyshev.**

$$
P\left(\left\vert \hat p - p\right\vert \ge \varepsilon\right) \le \frac{p(1-p)}{n\varepsilon^2} \le \frac{1}{4n\varepsilon^2} \le \delta \iff n \ge \frac{1}{4\delta\varepsilon^2} = \frac{1}{4(0.05)(0.0009)} \approx 5556 .
$$

**(b) Hoeffding.** With summands in $[0,1]$,

$$
2\exp\left(-2n\varepsilon^2\right) \le \delta \iff n \ge \frac{\ln(2/\delta)}{2\varepsilon^2} = \frac{\ln 40}{2(0.0009)} \approx \frac{3.689}{0.0018} \approx 2050 .
$$

**(c) CLT.** Requiring $z_{1-\delta/2}\sqrt{p(1-p)/n} \le \varepsilon$ with $z_{0.975} = 1.96$:

$$
n \ge \frac{z^2 p(1-p)}{\varepsilon^2} \le \frac{(1.96)^2 (0.25)}{0.0009} \approx 1068 .
$$

**Interpretation.** Chebyshev's dependence on $1/\delta$ is the expensive part; Hoeffding replaces it with $\ln(1/\delta)$ and is still a rigorous guarantee; the CLT number is an approximation that happens to be closest to the truth and is the one used in practice (the classic "about 1000 respondents for $\pm 3\%$").

$$
\boxed{n_{\text{Cheb}} \approx 5556, \quad n_{\text{Hoeff}} \approx 2050, \quad n_{\text{CLT}} \approx 1068}
$$

*Key takeaway:* Guarantees cost sample size; the price of rigor is the difference between $1/\delta$, $\log(1/\delta)$, and a Gaussian quantile.

Recompute the three sample sizes, taking ceilings as a pollster would.

In [3]:
eps_p, delta_p = 0.03, 0.05
n_cheb = np.ceil(1 / (4 * delta_p * eps_p**2))
n_hoeff = np.ceil(np.log(2 / delta_p) / (2 * eps_p**2))
n_clt = np.ceil(norm.ppf(1 - delta_p / 2) ** 2 * 0.25 / eps_p**2)
print(f"Chebyshev  n >= {n_cheb:.0f}")
print(f"Hoeffding  n >= {n_hoeff:.0f}")
print(f"CLT        n >= {n_clt:.0f}")
print(f"ratios to the CLT answer: {n_cheb / n_clt:.2f}x and {n_hoeff / n_clt:.2f}x")
assert (n_cheb, n_hoeff, n_clt) == (5556.0, 2050.0, 1068.0)

Chebyshev  n >= 5556
Hoeffding  n >= 2050
CLT        n >= 1068
ratios to the CLT answer: 5.20x and 1.92x


The boxed numbers $5556$, $2050$ and $1068$ are reproduced exactly. Rigour costs a factor
$5.2$ over the CLT approximation if bought with Chebyshev and only $1.9$ if bought with
Hoeffding — the whole difference is $1/\delta$ against $\ln(1/\delta)$.

### Problem L1.2 — Variance of the Sample Mean under Correlation

Let $X_1, \ldots, X_n$ be identically distributed with variance $\sigma^2$ and equicorrelation $\operatorname{Corr}(X_i, X_j) = \rho$ for $i \ne j$. Compute $\operatorname{Var}(\bar{X}_n)$ and determine the limit as $n \to \infty$. What does this say about the LLN?

**Intuition.** Correlation adds $n(n-1)$ covariance terms against only $n$ variance terms, so for large $n$ the covariances dominate and the shared component never averages away.

**Solution**

Expand the variance of the sum with all covariance terms retained:

$$
\operatorname{Var}\left(\sum_{i=1}^n X_i\right) = \sum_{i=1}^n \operatorname{Var}(X_i) + \sum_{i \ne j}\operatorname{Cov}(X_i, X_j) = n\sigma^2 + n(n-1)\rho\sigma^2 .
$$

Dividing by $n^2$:

$$
\operatorname{Var}(\bar{X}_n) = \frac{\sigma^2}{n}\left[1 + (n-1)\rho\right] = \sigma^2\left(\rho + \frac{1-\rho}{n}\right) .
$$

**Limit.** As $n \to \infty$, $\operatorname{Var}(\bar{X}_n) \to \rho\sigma^2$. For $\rho \gt 0$ this is a *positive constant*: averaging never eliminates the common component, and the LLN fails — the sample mean converges to a random limit, not to $\mu$. Only $\rho = 0$ recovers $\sigma^2/n$. (Positive semi-definiteness also forces $\rho \ge -1/(n-1)$, so strong negative equicorrelation is impossible for large $n$.)

**Effective sample size.** Matching $\sigma^2/n_{\text{eff}}$ to the true variance gives

$$
n_{\text{eff}} = \frac{n}{1 + (n-1)\rho} \xrightarrow{n\to\infty} \frac{1}{\rho} .
$$

Ten thousand highly correlated MCMC draws with $\rho = 0.9$ carry about the information of $\sim 1.1$ independent draws in the equicorrelated idealization; in practice one uses the integrated autocorrelation time.

$$
\boxed{\operatorname{Var}(\bar{X}_n) = \sigma^2\left(\rho + \frac{1-\rho}{n}\right) \to \rho\sigma^2}
$$

*Key takeaway:* Independence is not decoration — correlated data have an information ceiling that more samples cannot break.

Evaluate the variance formula and the effective sample size at $\rho = 0.9$.

In [4]:
sigma2_c, rho_c = 1.0, 0.9
ns_c = np.array([10, 100, 1_000, 10_000])
var_bar = sigma2_c * (rho_c + (1 - rho_c) / ns_c)
n_eff = ns_c / (1 + (ns_c - 1) * rho_c)
for n_c, v, ne in zip(ns_c, var_bar, n_eff):
    print(f"n = {n_c:6d}   Var(Xbar_n) = {v:.5f}   n_eff = {ne:.4f}")
print(f"limits as n -> infinity:  Var -> {rho_c * sigma2_c:.4f},  n_eff -> {1 / rho_c:.4f}")
assert abs(var_bar[-1] - 0.90001) < 1e-5
assert abs(n_eff[-1] - 1.1111) < 1e-4

n =     10   Var(Xbar_n) = 0.91000   n_eff = 1.0989
n =    100   Var(Xbar_n) = 0.90100   n_eff = 1.1099
n =   1000   Var(Xbar_n) = 0.90010   n_eff = 1.1110
n =  10000   Var(Xbar_n) = 0.90001   n_eff = 1.1111
limits as n -> infinity:  Var -> 0.9000,  n_eff -> 1.1111


The variance of the sample mean flattens at $\rho\sigma^2 = 0.9$ instead of falling to zero, and
ten thousand equicorrelated draws carry the information of $1.11$ independent ones. More data
cannot break an information ceiling set by the shared component.

### Problem L1.3 — Normal Approximation to the Binomial with Continuity Correction

Let $X \sim \text{Bin}(400, 0.25)$. Estimate $P(X \le 90)$ using the CLT with and without a continuity correction, and comment on the accuracy (exact value $\approx 0.13585$).

**Intuition.** A lattice variable is being approximated by a continuous one, so the natural cut point sits halfway between two integers rather than on one of them.

**Solution**

**Moments.** $\mu = np = 100$, $\sigma^2 = np(1-p) = 400(0.25)(0.75) = 75$, so $\sigma = \sqrt{75} \approx 8.660$.

**Without correction.**

$$
P(X \le 90) \approx \Phi\!\left(\frac{90 - 100}{8.660}\right) = \Phi(-1.1547) \approx 0.1241 .
$$

**With continuity correction.** The discrete event $\{X \le 90\}$ corresponds to the continuous region up to $90.5$:

$$
P(X \le 90) \approx \Phi\!\left(\frac{90.5 - 100}{8.660}\right) = \Phi(-1.0970) \approx 0.1363 .
$$

**Accuracy.** The exact binomial value is $P(X \le 90) = 0.13585$ (scipy.stats.binom.cdf(90, 400, 0.25)). Errors are $0.0117$ (uncorrected) and $0.0005$ (corrected): the half-unit shift cuts the error by a factor of about $24$, not merely "roughly half" — the continuity correction removes almost all of the systematic bias from approximating a step function by a smooth one. The tiny residual error is the skewness term of the Edgeworth expansion,

$$
\frac{\gamma}{6\sqrt{n}}(1-x^2)\varphi(x), \qquad \gamma = \frac{1-2p}{\sqrt{p(1-p)}} = \frac{0.5}{0.4330} \approx 1.155 ,
$$

which is nonzero precisely because $p \ne 1/2$ makes the binomial skewed.

$$
\boxed{P(X \le 90) \approx 0.1363 \text{ (corrected)}, \quad 0.1241 \text{ (uncorrected)}, \quad 0.1358 \text{ (exact)}}
$$

*Key takeaway:* Always apply the $\pm \tfrac{1}{2}$ correction when approximating lattice distributions by a continuous one.

Verify the exact binomial value and both normal approximations numerically.

In [5]:
from scipy.stats import binom, norm

mu, sigma = 100.0, np.sqrt(75.0)
exact = binom.cdf(90, 400, 0.25)
uncorrected = norm.cdf((90 - mu) / sigma)
corrected = norm.cdf((90.5 - mu) / sigma)
print(f"exact={exact:.5f}  uncorrected={uncorrected:.5f} (err {abs(uncorrected-exact):.5f})"
      f"  corrected={corrected:.5f} (err {abs(corrected-exact):.5f})"
      f"  error reduction factor={abs(uncorrected-exact)/abs(corrected-exact):.1f}x")

exact=0.13585  uncorrected=0.12411 (err 0.01174)  corrected=0.13633 (err 0.00048)  error reduction factor=24.3x


The continuity correction cuts the error by roughly 24x, confirming the boxed values above.

### Problem L1.4 — CLT for a Sum of Exponentials versus the Exact Gamma Law

Let $X_1, \ldots, X_{50}$ be i.i.d. $\text{Exponential}(1)$ and $S = \sum_i X_i$. Compute the CLT approximation to $P(S \gt 60)$, compare it in structure with the exact Gamma tail, and explain the direction of the error.

**Intuition.** A sum of exponentials is right-skewed, so a symmetric Gaussian must place too little mass in the upper tail; the first Edgeworth term predicts the sign of that deficit and most of its size.

**Solution**

**Moments.** For $\text{Exponential}(1)$, $\mu = 1$ and $\sigma^2 = 1$, so $E[S] = 50$ and $\operatorname{Var}(S) = 50$, $\operatorname{sd}(S) = \sqrt{50} \approx 7.071$.

**CLT approximation.**

$$
P(S \gt 60) \approx 1 - \Phi\!\left(\frac{60 - 50}{7.071}\right) = 1 - \Phi(1.4142) \approx 0.0786 .
$$

**Exact law.** $S \sim \text{Gamma}(50, 1)$, whose survival function is the Erlang tail

$$
P(S \gt s) = e^{-s}\sum_{k=0}^{49}\frac{s^k}{k!} , \qquad P(S \gt 60) \approx 0.08441 .
$$

**Direction of the error.** The exponential law has skewness $\gamma = 2 \gt 0$, so the sum's law is right-skewed: it carries more mass in the upper tail than a symmetric Gaussian with the same first two moments. The CLT therefore *underestimates* the right tail (and overestimates the left). The Edgeworth correction predicts the sign and magnitude:

$$
P(Z_n \le x) \approx \Phi(x) + \frac{\gamma}{6\sqrt{n}}\left(1 - x^2\right)\varphi(x), \qquad \frac{2}{6\sqrt{50}}\left(1 - 2\right)\varphi(1.4142) = -0.0069 ,
$$

and since the correction enters the *tail* with the opposite sign, the corrected tail is $\approx 0.07865 + 0.00692 = 0.08557$. That *overshoots* the true value $0.08441$: the first Edgeworth term removes about $80\%$ of the CLT error and then slightly overcorrects rather than landing exactly on target.

$$
\boxed{P(S \gt 60): \ \text{CLT} \approx 0.0786, \ \text{Edgeworth} \approx 0.0856, \ \text{exact} \approx 0.0844}
$$

*Key takeaway:* For skewed summands the CLT error is systematic and one-sided in the tails; the first Edgeworth term removes most of it but can overshoot the true value.

Verify the exact Erlang tail against the CLT and Edgeworth approximations.

In [6]:
from scipy.stats import poisson

exact_tail = poisson.cdf(49, 60)  # P(S > 60) for S ~ Gamma(50,1), Erlang-Poisson duality
x_e = (60 - 50) / np.sqrt(50)
clt_approx = 1 - norm.cdf(x_e)
# the Edgeworth term corrects the CDF, so it enters the tail with the opposite sign
edge_term = (2 / (6 * np.sqrt(50))) * (1 - x_e**2) * norm.pdf(x_e)
edgeworth = clt_approx - edge_term
print(f"exact={exact_tail:.5f}  CLT={clt_approx:.5f}  "
      f"Edgeworth term={edge_term:.5f}  Edgeworth tail={edgeworth:.5f}")
print(f"CLT error={abs(clt_approx - exact_tail):.5f}  "
      f"Edgeworth error={abs(edgeworth - exact_tail):.5f}  "
      f"removed {100 * (1 - abs(edgeworth - exact_tail) / abs(clt_approx - exact_tail)):.0f}% "
      f"of the CLT error")
assert abs(exact_tail - 0.08441) < 5e-6
assert abs(edgeworth - 0.08557) < 5e-6

exact=0.08441  CLT=0.07865  Edgeworth term=-0.00692  Edgeworth tail=0.08557
CLT error=0.00576  Edgeworth error=0.00116  removed 80% of the CLT error


The exact Erlang tail is $0.08441$; the plain CLT gives $0.07865$ and the Edgeworth-corrected
value $0.08557$, which removes $80\%$ of the CLT error and overshoots the truth by $0.00116$.

### Problem L1.5 — The Delta Method in Action

Let $X_1, \ldots, X_n$ be i.i.d. $\text{Bernoulli}(p)$ with $\hat{p} = \bar{X}_n$. Find the asymptotic distribution of (a) the odds $\hat{p}/(1-\hat{p})$ and (b) the log-odds $\operatorname{logit}(\hat p) = \ln\frac{\hat p}{1-\hat p}$. Then find a transformation whose asymptotic variance does not depend on $p$.

**Intuition.** A smooth transform locally multiplies the deviation by $g'(\theta)$, so it multiplies the asymptotic variance by $g'(\theta)^2$; choosing $g$ to cancel the $p$-dependence is then a one-line differential equation.

**Solution**

**Base CLT.** $\sqrt{n}(\hat{p} - p) \xrightarrow{d} \mathcal{N}(0, p(1-p))$.

**(a) Odds.** With $g(p) = \frac{p}{1-p}$,

$$
g'(p) = \frac{(1-p) + p}{(1-p)^2} = \frac{1}{(1-p)^2} ,
$$

so by the delta method

$$
\sqrt{n}\left(\frac{\hat p}{1-\hat p} - \frac{p}{1-p}\right) \xrightarrow{d} \mathcal{N}\!\left(0, \frac{p(1-p)}{(1-p)^4}\right) = \mathcal{N}\!\left(0, \frac{p}{(1-p)^3}\right) .
$$

**(b) Log-odds.** With $h(p) = \ln p - \ln(1-p)$,

$$
h'(p) = \frac{1}{p} + \frac{1}{1-p} = \frac{1}{p(1-p)} ,
$$

hence

$$
\sqrt{n}\left(\operatorname{logit}(\hat p) - \operatorname{logit}(p)\right) \xrightarrow{d} \mathcal{N}\!\left(0, \frac{p(1-p)}{p^2(1-p)^2}\right) = \mathcal{N}\!\left(0, \frac{1}{p(1-p)}\right) .
$$

This is why logistic-regression coefficients have standard error $1/\sqrt{n\hat p(1-\hat p)}$ in the intercept-only model, and why logit intervals back-transformed to probabilities respect the $[0,1]$ range.

**(c) Variance stabilization.** Seek $g$ with $g'(p)^2 p(1-p) = 1$, i.e. $g'(p) = \left[p(1-p)\right]^{-1/2}$. Integrating,

$$
g(p) = 2\arcsin\sqrt{p} \quad \Longrightarrow \quad \sqrt{n}\left(2\arcsin\sqrt{\hat p} - 2\arcsin\sqrt{p}\right) \xrightarrow{d} \mathcal{N}(0, 1) .
$$

$$
\boxed{\operatorname{AVar}(\hat p) = \tfrac{p(1-p)}{n}, \quad \operatorname{AVar}(\operatorname{logit}\hat p) = \tfrac{1}{np(1-p)}, \quad \operatorname{AVar}(2\arcsin\sqrt{\hat p}) = \tfrac{1}{n}}
$$

*Key takeaway:* The delta method converts one CLT into a CLT for every smooth functional; choosing the transform can make the variance parameter-free.

### Problem L1.6 — Convergence in Probability Implies Convergence in Distribution

Prove that $X_n \xrightarrow{P} X$ implies $X_n \xrightarrow{d} X$, and show by counterexample that the converse fails. Then prove the partial converse when the limit is a constant.

**Intuition.** Convergence in probability pins the random variables themselves together, while convergence in distribution only compares CDFs — so it must be weaker, except when the limit is a constant and there is no randomness left to mismatch.

**Solution**

**Forward implication.** Fix a continuity point $x$ of $F_X$ and $\varepsilon \gt 0$. Split on whether $X_n$ and $X$ are close:

$$
\{X_n \le x\} \subseteq \{X \le x + \varepsilon\} \cup \{\vert X_n - X\vert \gt \varepsilon\} ,
$$

because if $X_n \le x$ and $\vert X_n - X\vert \le \varepsilon$ then $X \le x + \varepsilon$. Hence

$$
F_n(x) \le F_X(x+\varepsilon) + P\left(\vert X_n - X\vert \gt \varepsilon\right) .
$$

Symmetrically, $\{X \le x - \varepsilon\} \subseteq \{X_n \le x\} \cup \{\vert X_n - X\vert \gt \varepsilon\}$ gives

$$
F_X(x-\varepsilon) - P\left(\vert X_n - X\vert \gt \varepsilon\right) \le F_n(x) .
$$

Letting $n \to \infty$ kills the probability terms, so

$$
F_X(x - \varepsilon) \le \liminf_n F_n(x) \le \limsup_n F_n(x) \le F_X(x+\varepsilon) .
$$

Now let $\varepsilon \downarrow 0$; continuity of $F_X$ at $x$ squeezes both bounds to $F_X(x)$, giving $F_n(x) \to F_X(x)$. $\blacksquare$

**Counterexample to the converse.** Let $X \sim \mathcal{N}(0,1)$ and set $X_n = -X$ for every $n$. Then $X_n \sim \mathcal{N}(0,1)$ for all $n$, so trivially $X_n \xrightarrow{d} X$. But $\vert X_n - X\vert = 2\vert X\vert$, and $P(2\vert X\vert \gt \varepsilon) \to$ a positive constant, not $0$. Convergence in distribution ignores the joint coupling entirely.

**Partial converse (constant limit).** If $X_n \xrightarrow{d} c$, then $F_n \to \mathbf{1}_{[c,\infty)}$ at all points except $c$, so for $\varepsilon \gt 0$,

$$
P\left(\vert X_n - c\vert \gt \varepsilon\right) \le F_n(c - \varepsilon) + \left(1 - F_n(c + \varepsilon/2)\right) \to 0 + 0 = 0 . \qquad \blacksquare
$$

$$
\boxed{\text{a.s.} \Rightarrow P \Rightarrow d; \quad d \Rightarrow P \text{ only for constant limits}}
$$

*Key takeaway:* Distributional convergence is a statement about laws; probabilistic convergence is a statement about the random variables themselves.

## L2 — Applications (AI/ML and Physics)

### Problem L2.1 — Photon Counting: Shot Noise and the Exposure Budget

A CCD pixel images a faint star that liberates photoelectrons at a mean rate $\lambda = 500\ \mathrm{s}^{-1}$. Arrivals in disjoint time intervals are independent, so the count in an exposure of length $t$ is $N(t) \sim \operatorname{Poisson}(\lambda t)$. (a) Give the sampling law of the rate estimate $\hat\lambda = N(t)/t$ and its relative uncertainty. (b) How long must the exposure be for a $95\%$ interval of relative half-width $10^{-3}$? (c) The detector also produces dark counts at a known rate $d = 200\ \mathrm{s}^{-1}$ — how does the budget change? (d) Where does the Gaussian approximation break down?

**Intuition.** Split the exposure into $m$ tiny disjoint sub-intervals: the total count is a sum of $m$ i.i.d. counts, so the CLT applies as $m \to \infty$, and all of shot noise follows from the Poisson identity $\operatorname{Var} = \text{mean}$.

**Solution**

**Step 1 — the sampling law.** Poisson counts are additive over disjoint intervals, so $N(t) = \sum_{k=1}^{m} N_k$ with $N_k$ i.i.d. $\operatorname{Poisson}(\lambda t/m)$. Theorem 4.7 applies for large $m$, and since $E[N(t)] = \operatorname{Var}(N(t)) = \lambda t$,

$$
\hat\lambda = \frac{N(t)}{t}, \qquad E[\hat\lambda] = \lambda, \qquad \operatorname{Var}(\hat\lambda) = \frac{\lambda}{t}, \qquad \hat\lambda \ \overset{\text{approx}}{\sim}\ \mathcal{N}\!\left(\lambda, \frac{\lambda}{t}\right) .
$$

**Step 2 — shot noise.** The relative uncertainty is

$$
\frac{\operatorname{sd}(\hat\lambda)}{\lambda} = \frac{1}{\sqrt{\lambda t}} = \frac{1}{\sqrt{N}}, \qquad \text{i.e.}\quad \mathrm{SNR} = \sqrt{N} .
$$

Precision is fixed by the number of photons actually collected, and by nothing else: no binning, gain or readout scheme changes it.

**Step 3 — the exposure budget.** A $95\%$ interval of relative half-width $\epsilon = 10^{-3}$ requires $1.96/\sqrt{\lambda t} \le \epsilon$, i.e.

$$
\lambda t \ge \frac{1.96^2}{\epsilon^2} = 3.8416\times 10^{6}, \qquad t \ge \frac{3.8416\times 10^{6}}{500} = 7683.2\ \mathrm{s} \approx 2.13\ \text{hours} .
$$

Nearly four million photoelectrons are needed for three digits — the $n^{-1/2}$ law in observational form: one more digit costs $100\times$ the telescope time.

**Step 4 — dark current.** The recorded total is $M(t) \sim \operatorname{Poisson}((\lambda + d)t)$, and since $d$ is known the estimator $\hat\lambda = M(t)/t - d$ is unbiased with $\operatorname{Var}(\hat\lambda) = (\lambda + d)/t$. Subtracting a known background removes the bias but keeps its variance. The absolute target is $\epsilon\lambda = 0.5\ \mathrm{s}^{-1}$, so

$$
t \ge \frac{1.96^2(\lambda + d)}{(\epsilon\lambda)^2} = \frac{3.8416 \times 700}{0.25} = 10756.5\ \mathrm{s} \approx 2.99\ \text{hours},
$$

a factor $(\lambda + d)/\lambda = 1.4$ longer. This is the whole reason detectors are cooled.

**Step 5 — where the Gaussian fails.** The skewness of $\operatorname{Poisson}(m)$ is $m^{-1/2}$, so the Edgeworth correction of Proof 5.6 is $O(m^{-1/2})$. At $m = \lambda t \approx 3.8\times 10^{6}$ it is $5\times 10^{-4}$ and irrelevant. In the photon-starved regime $m \lesssim 10$ — faint sources, short exposures, single-photon detectors — the count law is visibly skewed and discrete, the symmetric interval $\hat\lambda \pm 1.96\sqrt{\hat\lambda/t}$ under-covers badly, and exact Poisson (Garwood) intervals are required instead.

$$
\boxed{\hat\lambda \ \overset{\text{approx}}{\sim}\ \mathcal{N}\!\left(\lambda, \tfrac{\lambda}{t}\right), \qquad t_{\text{ideal}} = 7683\ \mathrm{s}, \qquad t_{\text{dark}} = 10757\ \mathrm{s} = 1.4\,t_{\text{ideal}}}
$$

*Key takeaway:* Shot noise is the CLT applied to photon arrivals — relative precision is $1/\sqrt{N}$, background counts inflate the exposure by $(\lambda+d)/\lambda$, and the Gaussian interval is only trustworthy once tens of counts have accumulated.

Recompute both exposure budgets, check $\operatorname{sd}(\hat\lambda) = \sqrt{\lambda/t}$ and the coverage of the Gaussian interval, then watch that coverage collapse in the photon-starved regime.

In [7]:
lam, dark, eps_rel, z95 = 500.0, 200.0, 1e-3, 1.96
t_ideal = z95**2 / (lam * eps_rel**2)
t_dark = z95**2 * (lam + dark) / (eps_rel * lam) ** 2
print(f"t_ideal = {t_ideal:8.1f} s = {t_ideal / 3600:.2f} h   photoelectrons = {lam * t_ideal:.4e}")
print(f"t_dark  = {t_dark:8.1f} s = {t_dark / 3600:.2f} h   ratio = {t_dark / t_ideal:.3f}"
      f"   (lam+d)/lam = {(lam + dark) / lam:.3f}")

t_probe = 100.0
lam_hat = rng.poisson(lam * t_probe, 200_000) / t_probe
rel_sd = float(lam_hat.std(ddof=1) / lam)
half = z95 * np.sqrt(lam_hat / t_probe)
print(f"\nt = {t_probe:.0f} s: measured relative sd = {rel_sd:.5f}   "
      f"predicted 1/sqrt(lam t) = {1 / np.sqrt(lam * t_probe):.5f}")
print(f"coverage of the 95% Gaussian interval = {np.mean(np.abs(lam_hat - lam) <= half):.4f}")

print("\nphoton-starved regime:")
for t_small in [0.006, 0.02, 0.2]:
    c = rng.poisson(lam * t_small, 200_000) / t_small
    hw = z95 * np.sqrt(c / t_small)
    print(f"  mean count = {lam * t_small:5.1f}   skewness 1/sqrt(m) = {1 / np.sqrt(lam * t_small):.3f}"
          f"   coverage = {np.mean(np.abs(c - lam) <= hw):.4f}")

assert abs(t_ideal - 7683.2) < 0.5 and abs(t_dark - 10756.5) < 0.5
assert abs(rel_sd - 1 / np.sqrt(lam * t_probe)) < 2e-4

t_ideal =   7683.2 s = 2.13 h   photoelectrons = 3.8416e+06
t_dark  =  10756.5 s = 2.99 h   ratio = 1.400   (lam+d)/lam = 1.400

t = 100 s: measured relative sd = 0.00447   predicted 1/sqrt(lam t) = 0.00447
coverage of the 95% Gaussian interval = 0.9501

photon-starved regime:
  mean count =   3.0   skewness 1/sqrt(m) = 0.577   coverage = 0.7968


  mean count =  10.0   skewness 1/sqrt(m) = 0.316   coverage = 0.9260


  mean count = 100.0   skewness 1/sqrt(m) = 0.100   coverage = 0.9452


Both boxed budgets reproduce, and the $1.4\times$ dark-current penalty is exactly $(\lambda+d)/\lambda$.
At $t = 100\ \mathrm{s}$ the measured relative spread matches $1/\sqrt{\lambda t} = 0.00447$ and the
Gaussian interval covers at $0.950$. In the photon-starved rows the same interval covers only
$0.797$ at a mean of $3$ counts and $0.926$ at $10$ — the Gaussian approximation fails exactly where
the Poisson skewness $m^{-1/2}$ stops being small, which is the finite-sample warning of
Theorem 4.10, not a defect of the simulation.

### Problem L2.2 — Minibatch Gradient Noise and the Linear Scaling Rule

In SGD the minibatch gradient is $\hat{g}_B = \frac{1}{B}\sum_{i \in \mathcal{B}}\nabla \ell_i(\theta)$ with i.i.d. sampling from a dataset whose per-example gradient covariance is $\Sigma$. Derive the distribution of the parameter update noise and explain the empirical "linear scaling rule" (double the batch, double the learning rate).

**Intuition.** A minibatch gradient is a sample mean, so its covariance is $\Sigma/B$; the learning rate then multiplies that noise, and holding the product fixed is what keeps two training runs statistically comparable.

**Solution**

**Unbiasedness and covariance.** Sampling examples uniformly with replacement makes each $\nabla \ell_i$ an i.i.d. draw with mean $g = \nabla L(\theta)$ and covariance $\Sigma$. Hence

$$
E[\hat g_B] = g, \qquad \operatorname{Cov}(\hat g_B) = \frac{\Sigma}{B} .
$$

By the CLT, for moderately large $B$,

$$
\hat{g}_B \approx g + \frac{1}{\sqrt{B}}\Sigma^{1/2}\xi, \qquad \xi \sim \mathcal{N}(0, I) .
$$

**Update noise.** One SGD step with learning rate $\eta$ is $\Delta\theta = -\eta \hat g_B$, so

$$
E[\Delta\theta] = -\eta g, \qquad \operatorname{Cov}(\Delta\theta) = \frac{\eta^2}{B}\Sigma .
$$

**Noise per unit progress.** Over a fixed number of *epochs*, the number of steps is $\propto 1/B$. Accumulating $k = T/B$ independent steps, the drift is $-\eta k g \propto \eta T g/B$ and the accumulated noise covariance is $k \eta^2 \Sigma/B \propto \eta^2 T\Sigma/B^2$. Matching both the drift per epoch and the noise scale across batch sizes requires

$$
\frac{\eta}{B} = \text{const} \quad \Longrightarrow \quad \eta \propto B ,
$$

which is exactly the linear scaling rule. Equivalently, the continuous-time limit of SGD is the SDE

$$
d\theta = -\nabla L(\theta)\,dt + \sqrt{\frac{\eta}{B}}\,\Sigma^{1/2}\,dW ,
$$

whose stationary distribution depends on $\theta$ only through the *temperature* $\eta/B$. Two configurations with equal $\eta/B$ explore the loss landscape identically — the basis for the "large batch needs large learning rate (with warm-up)" practice.

**Physical reading.** That SDE is the overdamped Langevin equation of a Brownian particle in the potential $L(\theta)$: $-\nabla L$ is the deterministic force, $\Sigma^{1/2}\,dW$ is the thermal kick, and $\eta/B$ plays the role of $k_BT$. For isotropic noise $\Sigma = \sigma^2 I$ the diffusion coefficient is $D = \eta\sigma^2/(2B)$ and the stationary density is the Gibbs law

$$
p_\infty(\theta) \propto \exp\left(-\frac{L(\theta)}{D}\right) = \exp\left(-\frac{2B\,L(\theta)}{\eta\sigma^2}\right) ,
$$

so raising the temperature $\eta/B$ broadens the occupied region exactly as heating a colloidal particle does, and lowering it anneals the iterate into the nearest sharp minimum. The Einstein relation $\langle \lVert \theta - \theta_0\rVert^2\rangle = 2Dt$ of free diffusion is recovered wherever the loss is locally flat.

$$
\boxed{\operatorname{Cov}(\hat g_B) = \Sigma/B; \quad \text{SGD temperature} = \eta/B \text{ must be held fixed}}
$$

*Key takeaway:* The CLT turns minibatching into calibrated Gaussian noise whose strength is $\eta/B$ — the single knob controlling SGD's implicit regularization.

Check the covariance law $\operatorname{Cov}(\hat g_B) = \Sigma/B$ on a quadratic loss, and
check that two runs with the same temperature $\eta/B$ inject the same noise per step.

In [8]:
d_g, N_g = 4, 3000
A_g = rng.standard_normal((N_g, d_g))
b_g = A_g @ rng.standard_normal(d_g) + 0.3 * rng.standard_normal(N_g)
theta_g = rng.standard_normal(d_g)
per_ex = A_g * (A_g @ theta_g - b_g)[:, None]
Sigma_g = np.cov(per_ex.T, ddof=1)

print("  B     tr Cov(g_B) measured   tr(Sigma)/B predicted   eta   eta^2 tr(Sigma)/B")
for B, eta in [(1, 0.01), (4, 0.04), (16, 0.16), (64, 0.64)]:
    idx = rng.integers(0, N_g, size=(4000, B))
    meas = float(np.trace(np.cov(per_ex[idx].mean(axis=1).T, ddof=1)))
    pred = float(np.trace(Sigma_g)) / B
    print(f"{B:4d}      {meas:14.5f}       {pred:14.5f}   {eta:.2f}   "
          f"{eta**2 * pred:.6f}")
    assert abs(meas - pred) / pred < 0.10

  B     tr Cov(g_B) measured   tr(Sigma)/B predicted   eta   eta^2 tr(Sigma)/B
   1            14.92301             14.31610   0.01   0.001432
   4             3.48916              3.57902   0.04   0.005726


  16             0.89995              0.89476   0.16   0.022906


  64             0.22708              0.22369   0.64   0.091623


The measured minibatch covariance matches $\Sigma/B$ to within a few percent at every batch
size. The last column holds the per-step update-noise scale $\eta^2\operatorname{tr}(\Sigma)/B$
along the line $\eta \propto B$: it grows like $\eta^2/B \propto B$, which is exactly why the
linear scaling rule must be paired with a matching reduction in the number of steps per epoch.

### Problem L2.3 — A/B Test Sizing for a Model Deployment

Model A has click-through rate $p_A = 0.10$. You want $80\%$ power to detect a lift to $p_B = 0.11$ at two-sided significance $\alpha = 0.05$. Derive the required per-arm sample size using the CLT.

**Intuition.** A power calculation says the true effect must clear both the null's critical value and the alternative's own spread, which is why the two Gaussian quantiles add before being squared.

**Solution**

**Test statistic.** With $\hat p_A, \hat p_B$ independent sample proportions from $n$ users each, the CLT gives

$$
\hat p_B - \hat p_A \approx \mathcal{N}\!\left(\delta, \ \frac{p_A(1-p_A) + p_B(1-p_B)}{n}\right), \qquad \delta = p_B - p_A = 0.01 .
$$

**Power condition.** Rejecting when $\vert \hat p_B - \hat p_A\vert \gt z_{1-\alpha/2}\,\text{se}_0$ and requiring power $1-\beta$ under the alternative gives the standard requirement

$$
\delta \ \ge\ \left(z_{1-\alpha/2} + z_{1-\beta}\right)\sqrt{\frac{p_A(1-p_A)+p_B(1-p_B)}{n}} ,
$$

so

$$
n \ \ge\ \frac{\left(z_{1-\alpha/2} + z_{1-\beta}\right)^2\left[p_A(1-p_A) + p_B(1-p_B)\right]}{\delta^2} .
$$

**Numbers.** $z_{0.975} = 1.96$, $z_{0.80} = 0.8416$, sum $= 2.8016$, squared $= 7.849$. Variances: $0.10(0.90) = 0.09$ and $0.11(0.89) = 0.0979$, total $0.1879$. Thus

$$
n \ge \frac{7.849 \times 0.1879}{10^{-4}} \approx 14{,}748 \ \text{per arm} \approx 29{,}500 \text{ total} .
$$

**Sensitivity.** Because $n \propto \delta^{-2}$, halving the detectable lift to $0.005$ quadruples the cost to $\approx 59{,}000$ per arm. This inverse-square wall is why teams switch to variance-reduced metrics (CUPED), stratification, or sequential tests rather than simply waiting longer.

$$
\boxed{n \approx 1.5\times 10^4 \text{ per arm for a } 10\% \to 11\% \text{ lift at } 80\% \text{ power}}
$$

*Key takeaway:* Experiment cost scales as (variance)/(effect size)$^2$ — shrinking metric variance is worth as much as doubling traffic.

Recompute the per-arm sample size from the two Gaussian quantiles.

In [9]:
z_alpha, z_beta = norm.ppf(0.975), norm.ppf(0.80)
pA, pB = 0.10, 0.11
var_sum = pA * (1 - pA) + pB * (1 - pB)
n_arm = (z_alpha + z_beta) ** 2 * var_sum / (pB - pA) ** 2
print(f"z_0.975 = {z_alpha:.4f}   z_0.80 = {z_beta:.4f}   (sum)^2 = {(z_alpha + z_beta)**2:.4f}")
print(f"variance sum = {var_sum:.4f}")
print(f"raw requirement n >= {n_arm:.1f}  ->  per-arm n = {np.ceil(n_arm):.0f}   "
      f"total = {2 * np.ceil(n_arm):.0f}")
print(f"halving the lift to 0.005 needs 4x that: {np.ceil(4 * n_arm):.0f} per arm")
assert abs(n_arm - 14748) < 2

z_0.975 = 1.9600   z_0.80 = 0.8416   (sum)^2 = 7.8489
variance sum = 0.1879
raw requirement n >= 14748.0  ->  per-arm n = 14749   total = 29498
halving the lift to 0.005 needs 4x that: 58993 per arm


The raw requirement is $14{,}748.0$, i.e. $14{,}749$ users per arm after rounding up, confirming
the boxed figure of about $1.5\times 10^4$. The inverse-square
dependence on the effect size is explicit: detecting half the lift costs four times the
traffic, roughly $59{,}000$ per arm.

### Problem L2.4 — Random Walk to Diffusion — Deriving the Heat Kernel

A particle takes $n$ independent steps of length $\pm \ell$ with equal probability every $\tau$ seconds. Use the CLT to derive the position distribution at time $t$ and identify the diffusion coefficient.

**Intuition.** A walker's position is a sum of i.i.d. steps, so the CLT supplies the Gaussian shape for free and the only physics left is matching the variance to elapsed time.

**Solution**

**Step statistics.** Each step $S_i$ satisfies $E[S_i] = 0$ and $E[S_i^2] = \ell^2$, so $\sigma^2 = \ell^2$. After $n = t/\tau$ steps the displacement is $X_n = \sum_{i=1}^n S_i$ with

$$
E[X_n] = 0, \qquad \operatorname{Var}(X_n) = n\ell^2 = \frac{\ell^2}{\tau}\,t .
$$

**CLT limit.** Since the steps are i.i.d. with finite variance,

$$
\frac{X_n}{\ell\sqrt{n}} \xrightarrow{d} \mathcal{N}(0,1) \quad \Longrightarrow \quad X(t) \approx \mathcal{N}\!\left(0, \frac{\ell^2}{\tau}t\right) .
$$

Writing the variance as $2Dt$ identifies the diffusion coefficient

$$
D = \frac{\ell^2}{2\tau} ,
$$

and the density is the heat kernel

$$
p(x, t) = \frac{1}{\sqrt{4\pi D t}}\exp\left(-\frac{x^2}{4Dt}\right) ,
$$

which indeed solves $\partial_t p = D\,\partial_x^2 p$.

**Physical reading.** Two consequences follow immediately. First, $\langle x^2\rangle = 2Dt$: diffusive spreading grows as $\sqrt{t}$, not $t$ — the microscopic $\sqrt{n}$ of the CLT becomes the macroscopic $\sqrt{t}$ of diffusion. Second, the limit shape is Gaussian regardless of the step distribution (any zero-mean finite-variance step law gives the same kernel with $D = \sigma^2/(2\tau)$), which is why Fick's law describes systems with wildly different microphysics. Heavy-tailed step lengths (Lévy flights) break this and yield fractional diffusion with $\langle \vert x\vert^\alpha\rangle \propto t$.

$$
\boxed{X(t) \sim \mathcal{N}(0, 2Dt), \quad D = \frac{\ell^2}{2\tau}, \quad \langle x^2\rangle = 2Dt}
$$

*Key takeaway:* Diffusion is the CLT in continuous time; universality of the Gaussian is why one coefficient $D$ summarizes all microscopic detail.

Simulate the walk and check both the diffusion coefficient and the Gaussian shape.

In [10]:
from scipy.stats import kstest

ell, tau, t_end = 0.5, 0.01, 2.0
n_steps = int(t_end / tau)
walks = rng.choice([-ell, ell], size=(40_000, n_steps)).sum(axis=1)
D_pred = ell**2 / (2 * tau)

print(f"n = t/tau = {n_steps} steps,  D = l^2/(2 tau) = {D_pred:.4f}")
print(f"simulated Var(X(t)) = {walks.var(ddof=1):.4f}   predicted 2Dt = {2 * D_pred * t_end:.4f}")
ks_stat = float(kstest(walks / np.sqrt(2 * D_pred * t_end), "norm").statistic)
print(f"KS distance from the simulated law to N(0, 2Dt) = {ks_stat:.4f}")
print(f"lattice spacing 2l = {2 * ell:.1f} against sd = {np.sqrt(2 * D_pred * t_end):.3f}")
assert abs(walks.var(ddof=1) - 2 * D_pred * t_end) / (2 * D_pred * t_end) < 0.03
assert ks_stat < 0.05

n = t/tau = 200 steps,  D = l^2/(2 tau) = 12.5000
simulated Var(X(t)) = 49.7716   predicted 2Dt = 50.0000


KS distance from the simulated law to N(0, 2Dt) = 0.0307
lattice spacing 2l = 1.0 against sd = 7.071


The simulated displacement variance matches $2Dt = 50$ with $D = \ell^2/(2\tau) = 12.5$, and the
standardized displacement sits within $0.031$ in Kolmogorov distance of the standard normal.
The residual is not sampling error: the walk lives on a lattice of spacing $2\ell = 1$ against a
standard deviation of $7.07$, and a step function cannot match a smooth CDF more closely than
half a lattice step. The heat kernel is the CLT limit of the walk, not an independent physical
assumption.

### Problem L2.5 — Why He Initialization Keeps Deep Networks Alive

A fully connected layer computes $z_j = \sum_{k=1}^{d} W_{jk} x_k$ with i.i.d. weights of mean $0$ and variance $\sigma_W^2$, independent of the inputs with $E[x_k^2] = q$. Determine the distribution of $z_j$ for large $d$, and derive the variance scaling that keeps signal magnitude constant through a ReLU network.

**Intuition.** A pre-activation is a sum of $d$ independent products, so the CLT makes it Gaussian; keeping a deep network alive is then just the requirement that this Gaussian's variance neither grow nor decay with depth.

**Solution**

**Pre-activation law.** The terms $W_{jk}x_k$ are i.i.d. (conditionally on $x$), zero-mean, with variance $\sigma_W^2 q$. By the CLT, for large fan-in $d$,

$$
z_j \approx \mathcal{N}\!\left(0,\ d\,\sigma_W^2 q\right) .
$$

This is why pre-activations at initialization are empirically Gaussian even when the inputs and weights are not — the layer performs the sum that the CLT governs.

**Propagating through ReLU.** Let $q^{(l)} = E[(x^{(l)})^2]$ denote the second moment of layer-$l$ activations. With $z \sim \mathcal{N}(0, v)$ and $x = \max(z, 0)$, symmetry gives

$$
E\left[\max(z,0)^2\right] = \frac{1}{2}E[z^2] = \frac{v}{2} .
$$

Therefore

$$
q^{(l+1)} = \frac{1}{2}\,d_l\,\sigma_W^2\,q^{(l)} .
$$

**Fixed point.** Signal neither explodes nor vanishes across depth iff the multiplier is $1$:

$$
\frac{d_l \sigma_W^2}{2} = 1 \quad \Longrightarrow \quad \sigma_W^2 = \frac{2}{d_l} ,
$$

which is exactly He (Kaiming) initialization. For $\tanh$-like activations with $E[\phi(z)^2] \approx E[z^2]$ near the origin the factor $1/2$ disappears and one recovers Glorot/Xavier scaling $\sigma_W^2 = 1/d_l$ (or $2/(d_{\text{in}} + d_{\text{out}})$ symmetrized for backward passes).

**Why it matters.** With $\sigma_W^2 = c/d$, the layer multiplier is $c/2$ per layer, so after $L$ layers the signal scales as $(c/2)^L$ — exponential in depth. A $10\%$ error in $c$ compounds to a factor $\approx 2.6$ over $L = 10$ and $\approx 1.3\times 10^{4}$ over $L = 100$, which is precisely the vanishing/exploding activation problem.

$$
\boxed{z \approx \mathcal{N}(0, d\sigma_W^2 q); \quad \text{ReLU stability} \iff \sigma_W^2 = 2/d_{\text{in}}}
$$

*Key takeaway:* Initialization schemes are CLT variance bookkeeping — they set the one constant that stops depth from exponentiating.

Propagate a signal through a $30$-layer ReLU stack at three weight scales and compare
the mean square activation with the predicted $(c/2)^{L}$.

In [11]:
d_layer, depth = 512, 30
x0 = rng.standard_normal((1024, d_layer))
x0 /= np.sqrt(np.mean(x0**2))                     # start with second moment 1

print("  c    q after 30 layers   predicted (c/2)^30   fitted per-layer factor   c/2")
for c in [1.0, 2.0, 4.0]:
    a = x0
    for _ in range(depth):
        W = np.sqrt(c / d_layer) * rng.standard_normal((d_layer, d_layer))
        a = np.maximum(a @ W, 0.0)
    q_final = float(np.mean(a**2))
    fitted = q_final ** (1 / depth)                # geometric mean multiplier per layer
    print(f"{c:4.1f}   {q_final:17.4e}   {(c / 2) ** depth:17.4e}   "
          f"{fitted:21.4f}   {c / 2:.2f}")
    assert abs(fitted - c / 2) / (c / 2) < 0.05

  c    q after 30 layers   predicted (c/2)^30   fitted per-layer factor   c/2


 1.0          1.0364e-09          9.3132e-10                  0.5018   0.50


 2.0          6.6852e-01          1.0000e+00                  0.9867   1.00


 4.0          5.6206e+08          1.0737e+09                  1.9573   2.00


Only $c = 2$ — He initialization — keeps the signal at order one after $30$ layers. With
$c = 1$ the mean square activation collapses to $10^{-9}$ and with $c = 4$ it explodes past
$10^{9}$: nineteen orders of magnitude separate the two, from a factor-of-four change in one
constant. A single random initialization scatters lognormally about $(c/2)^{30}$, so the sharp
test is the fitted per-layer multiplier $q_L^{1/L}$, which reproduces $c/2$ to within $5\%$ in
all three runs. That per-layer factor is exactly the CLT variance bookkeeping, and depth
exponentiates it.

### Problem L2.6 — Control Variates for a Noisy Estimator

You estimate $\theta = E[h(X)]$ by Monte Carlo but also know $E[c(X)] = \mu_c$ exactly for a correlated function $c$. Derive the optimal control-variate estimator, its variance, and the resulting reduction factor. Specialize to antithetic variates, and apply the idea to reinforcement-learning baselines.

**Intuition.** Subtracting a mean-zero quantity never changes an expectation but does change a variance, and the best multiple to subtract is exactly the least-squares regression slope.

**Solution**

**Estimator family.** For any constant $\beta$, define

$$
\hat\theta_\beta = \frac{1}{n}\sum_{i=1}^n \left[h(X_i) - \beta\left(c(X_i) - \mu_c\right)\right] .
$$

Since $E[c(X) - \mu_c] = 0$, every member is unbiased: $E[\hat\theta_\beta] = \theta$.

**Variance.** Per sample,

$$
\operatorname{Var}\left(h - \beta(c-\mu_c)\right) = \sigma_h^2 - 2\beta\operatorname{Cov}(h,c) + \beta^2\sigma_c^2 .
$$

Minimize over $\beta$ by setting the derivative to zero:

$$
-2\operatorname{Cov}(h,c) + 2\beta\sigma_c^2 = 0 \quad \Longrightarrow \quad \beta^\star = \frac{\operatorname{Cov}(h,c)}{\sigma_c^2} ,
$$

which is exactly the least-squares regression slope of $h$ on $c$. Substituting back,

$$
\operatorname{Var}\left(\hat\theta_{\beta^\star}\right) = \frac{\sigma_h^2}{n}\left(1 - \rho^2\right), \qquad \rho = \operatorname{Corr}(h, c) .
$$

**Reduction factor.** The variance shrinks by $1 - \rho^2$: a control variate correlated at $\rho = 0.9$ removes $81\%$ of the variance (a $5.3\times$ effective sample-size gain), while $\rho = 0.5$ removes only $25\%$. Correlation, not cleverness, is the currency.

**Antithetic variates.** The classical antithetic estimator is this construction with $\beta$ frozen at $-1$ and $c$ a mirrored copy of $h$. For $U \sim \operatorname{Unif}(0,1)$ and monotone $h$, the pair average $\left(h(U) + h(1-U)\right)/2$ has variance $\frac{\sigma_h^2}{2}\left(1 + \rho_a\right)$ with $\rho_a = \operatorname{Corr}\left(h(U), h(1-U)\right) \lt 0$. For $h(x) = e^{-x^2}$ on $[0,1]$ we have $\sigma_h = 0.2010$ and $\rho_a = -0.9601$, so a $95\%$ interval of half-width $10^{-3}$ costs $n = (1.96\,\sigma_h/10^{-3})^2 \approx 1.55\times 10^{5}$ plain draws but only $n(1+\rho_a) \approx 6.2\times 10^{3}$ function evaluations in antithetic pairs — a $25$-fold saving at no extra cost per sample.

**RL baselines.** In policy-gradient methods the estimator $\hat g = \frac{1}{n}\sum_i \nabla_\theta \ln \pi_\theta(a_i \mid s_i)\,R_i$ has enormous variance. Subtracting a state-dependent baseline $b(s)$ gives $\left(R_i - b(s_i)\right)$; because $E\left[\nabla_\theta \ln\pi_\theta(a\mid s)\right] = 0$ for each fixed $s$, the estimator stays unbiased for any $b$. The optimal $b$ is the regression of returns on the score — approximated in practice by the value function $V(s)$, which is why actor-critic and advantage estimation $A = R - V(s)$ work.

$$
\boxed{\beta^\star = \frac{\operatorname{Cov}(h,c)}{\operatorname{Var}(c)}, \qquad \operatorname{Var}(\hat\theta_{\beta^\star}) = \frac{\sigma_h^2(1-\rho^2)}{n}}
$$

*Key takeaway:* Any zero-mean correlated quantity is free variance reduction; the achievable gain is exactly $1 - \rho^2$.

Check the $1-\rho^2$ law, fit the optimal $\beta$ on a concrete integrand, and measure the antithetic correlation.

In [12]:
for r in [0.3, 0.5, 0.9, 0.99]:
    print(f"rho = {r:4.2f}   variance factor 1 - rho^2 = {1 - r**2:.4f}   "
          f"effective sample-size gain = {1 / (1 - r**2):6.2f}x")

U = rng.uniform(0, 1, 200_000)
h_u, c_u = np.exp(-U**2), U                       # control variate with known mean 1/2
beta_star = float(np.cov(h_u, c_u, ddof=1)[0, 1] / np.var(c_u, ddof=1))
var_plain = float(np.var(h_u, ddof=1))
var_cv = float(np.var(h_u - beta_star * (c_u - 0.5), ddof=1))
rho_hc = float(np.corrcoef(h_u, c_u)[0, 1])
print(f"\nbeta* = {beta_star:.4f}   Var(plain) = {var_plain:.6f}   "
      f"Var(control variate) = {var_cv:.6f}")
print(f"ratio = {var_cv / var_plain:.4f}   1 - rho^2 = {1 - rho_hc**2:.4f}")
assert abs(var_cv / var_plain - (1 - rho_hc**2)) < 1e-3

u_a = rng.uniform(0, 1, 200_000)
h_a, h_mirror = np.exp(-u_a**2), np.exp(-(1 - u_a) ** 2)
rho_a = float(np.corrcoef(h_a, h_mirror)[0, 1])
sigma_h = float(np.std(h_a, ddof=1))
n_plain_mc = (1.96 * sigma_h / 1e-3) ** 2
print(f"\nantithetic: sigma_h = {sigma_h:.4f}   rho_a = {rho_a:.4f}   "
      f"n_plain = {n_plain_mc:.3e}   n_antithetic evaluations = {n_plain_mc * (1 + rho_a):.3e}")
assert rho_a < -0.95

rho = 0.30   variance factor 1 - rho^2 = 0.9100   effective sample-size gain =   1.10x
rho = 0.50   variance factor 1 - rho^2 = 0.7500   effective sample-size gain =   1.33x
rho = 0.90   variance factor 1 - rho^2 = 0.1900   effective sample-size gain =   5.26x
rho = 0.99   variance factor 1 - rho^2 = 0.0199   effective sample-size gain =  50.25x

beta* = -0.6880   Var(plain) = 0.040254   Var(control variate) = 0.000927
ratio = 0.0230   1 - rho^2 = 0.0230

antithetic: sigma_h = 0.2006   rho_a = -0.9602   n_plain = 1.546e+05   n_antithetic evaluations = 6.152e+03


The measured variance ratio equals $1-\rho^2$ to four decimals, confirming the boxed formula.
A control variate correlated at $0.9$ buys a $5.3\times$ effective sample size; at $0.5$ it buys
only $1.33\times$. The antithetic pair reaches $\rho_a = -0.96$, cutting $1.55\times10^{5}$ plain draws
to about $6.2\times10^{3}$ evaluations. Correlation, not cleverness, is the currency.

## L3 — Challenge Proofs

### Problem L3.1 — Berry–Esseen and the Rare-Event Sample Size

For i.i.d. $\text{Bernoulli}(p)$ summands with small $p$, compute the Berry–Esseen ratio $\rho/\sigma^3$ and derive how large $n$ must be for the normal approximation to be uniformly accurate to $\epsilon$. Reconcile with the classical rule $np \ge 10$.

**Intuition.** The Berry–Esseen ratio $\rho/\sigma^3$ is scale-free, and for a rare event the standard deviation $\sqrt{p(1-p)}$ shrinks faster than the third absolute moment, so the ratio blows up like $p^{-1/2}$.

**Solution**

**Third absolute central moment.** For $X \sim \text{Bernoulli}(p)$ with $\mu = p$:

$$
\rho = E\left\vert X - p\right\vert^3 = p(1-p)^3 + (1-p)p^3 = p(1-p)\left[(1-p)^2 + p^2\right] .
$$

Since $\sigma^3 = \left[p(1-p)\right]^{3/2}$, the Berry–Esseen ratio is

$$
\frac{\rho}{\sigma^3} = \frac{(1-p)^2 + p^2}{\sqrt{p(1-p)}} .
$$

**Small-$p$ asymptotics.** As $p \to 0$, the numerator tends to $1$ and the denominator to $\sqrt{p}$:

$$
\frac{\rho}{\sigma^3} \sim \frac{1}{\sqrt{p}} \longrightarrow \infty .
$$

**Sample size.** Theorem 4.10 of `first_principles.ipynb` requires

$$
\frac{C\rho}{\sigma^3\sqrt{n}} \le \epsilon \quad \Longrightarrow \quad n \ \ge\ \left(\frac{C}{\epsilon}\right)^2 \frac{\left[(1-p)^2+p^2\right]^2}{p(1-p)} \ \approx\ \frac{C^2}{\epsilon^2 p} \quad (p \ll 1) .
$$

So $n$ must grow like $1/p$, i.e. **$np$ must exceed a constant** depending only on the target accuracy. With $C = 0.4748$ and $\epsilon = 0.05$ this constant is $(0.4748/0.05)^2 \approx 90$; with the sharper practical behavior (Berry–Esseen is conservative) the working threshold drops to the familiar $np \gtrsim 10$, and symmetrically $n(1-p) \gtrsim 10$ by applying the same analysis to $1 - X$.

**Interpretation.** $np$ is the *expected number of successes*: normal approximation needs enough successes, not merely enough trials. When $np$ stays bounded as $n$ grows, the correct limit is not Gaussian at all but Poisson($\lambda = np$) — the law of rare events, a different limit theorem entirely.

$$
\boxed{n \gtrsim \frac{C^2}{\epsilon^2 p} \iff np \gtrsim \text{const}; \text{ otherwise the limit is Poisson, not Gaussian}}
$$

*Key takeaway:* The CLT's validity threshold is governed by a scale-free third moment, and for rare events that quantity blows up like $p^{-1/2}$.

Tabulate the Berry–Esseen ratio against $p^{-1/2}$ and evaluate the implied threshold on $np$.

In [13]:
p_vals = np.array([0.5, 0.2, 0.05, 0.01, 0.001])
ratio = ((1 - p_vals) ** 2 + p_vals**2) / np.sqrt(p_vals * (1 - p_vals))
print("    p     rho / sigma^3     1 / sqrt(p)")
for p_v, r_v in zip(p_vals, ratio):
    print(f"{p_v:7.3f}   {r_v:11.4f}   {1 / np.sqrt(p_v):13.4f}")

C_be, eps_be = 0.4748, 0.05
print(f"\n(C/eps)^2 = ({C_be}/{eps_be})^2 = {(C_be / eps_be) ** 2:.1f}")
print("so Berry-Esseen guarantees eps = 0.05 uniformly once np exceeds about "
      f"{(C_be / eps_be) ** 2:.0f}")
assert abs((C_be / eps_be) ** 2 - 90.2) < 0.2
assert np.all(np.diff(ratio) > 0)

    p     rho / sigma^3     1 / sqrt(p)
  0.500        1.0000          1.4142
  0.200        1.7000          2.2361
  0.050        4.1524          4.4721
  0.010        9.8514         10.0000
  0.001       31.5754         31.6228

(C/eps)^2 = (0.4748/0.05)^2 = 90.2
so Berry-Esseen guarantees eps = 0.05 uniformly once np exceeds about 90


The ratio $\rho/\sigma^3$ tracks $p^{-1/2}$ closely once $p \le 0.05$ and grows without bound as
$p \to 0$. The guaranteed threshold $np \gtrsim 90$ is conservative — Berry–Esseen is a
worst-case bound — which is why the working folklore settles near $np \gtrsim 10$.

### Problem L3.2 — The Cauchy Distribution Defeats Averaging

Let $X_1, X_2, \ldots$ be i.i.d. standard Cauchy with density $f(x) = \frac{1}{\pi(1+x^2)}$ and characteristic function $\varphi(t) = e^{-\vert t\vert}$. (a) Show $E\vert X_1\vert = \infty$. (b) Find the exact law of $\bar{X}_n$. (c) Show that no normalization $a_n$ makes $\bar{X}_n/a_n$ Gaussian, and identify what does converge.

**Intuition.** The Cauchy characteristic function $e^{-\vert t\vert}$ is reproduced exactly by averaging, so the sample mean cannot improve on a single observation no matter how much data is collected.

**Solution**

**(a) Non-integrability.** By symmetry,

$$
E\vert X_1\vert = \frac{2}{\pi}\int_0^\infty \frac{x}{1+x^2}\,dx = \frac{1}{\pi}\Big[\ln(1+x^2)\Big]_0^\infty = \infty .
$$

The mean is undefined, so neither LLN has a target to converge to.

**(b) Exact law of the average.** Using independence and the scaling rule $\varphi_{aX}(t) = \varphi_X(at)$,

$$
\varphi_{\bar{X}_n}(t) = \prod_{i=1}^n \varphi_{X_i}\!\left(\frac{t}{n}\right) = \left(e^{-\vert t\vert/n}\right)^{n} = e^{-\vert t\vert} = \varphi_{X_1}(t) .
$$

By uniqueness of characteristic functions, $\bar{X}_n \overset{d}{=} X_1$ for **every** $n$: averaging a million Cauchy observations gives a value distributed exactly like a single observation. The sample mean is worthless as an estimator; the sample *median*, by contrast, is asymptotically normal with variance $\pi^2/(4n)$ and is the practical location estimator.

**(c) No Gaussian limit.** Suppose $\bar{X}_n/a_n \xrightarrow{d} \mathcal{N}(0,1)$ for some $a_n \gt 0$. Then

$$
\varphi_{\bar{X}_n/a_n}(t) = e^{-\vert t\vert/a_n} \longrightarrow e^{-t^2/2} .
$$

At $t$ and $2t$ the left side satisfies $\varphi(2t) = \varphi(t)^2$ for all $n$, whereas the Gaussian satisfies $\varphi(2t) = \varphi(t)^4$; these are incompatible for $t \ne 0$. Concretely, if $a_n \to a \in (0,\infty)$ the limit is Cauchy; if $a_n \to \infty$ the limit is the constant $0$; if $a_n \to 0$ there is no proper limit. In no case is it Gaussian.

**What does converge.** The Cauchy law is $1$-stable: $S_n/n \overset{d}{=} X_1$, matching the general stable scaling $S_n/n^{1/\alpha}$ with $\alpha = 1$. The Gaussian is the special case $\alpha = 2$, available exactly when the variance is finite.

$$
\boxed{\bar{X}_n \overset{d}{=} X_1 \ \forall n; \ \text{Cauchy is } 1\text{-stable, outside the Gaussian domain of attraction}}
$$

*Key takeaway:* Finite variance is not a technicality — without it the sample mean can carry literally no more information than a single observation.

### Problem L3.3 — From Chernoff to Hoeffding

Derive Hoeffding's inequality for bounded i.i.d. summands: for $X_i \in [a,b]$ with mean $\mu$,

$$
P\left(\bar{X}_n - \mu \ge \varepsilon\right) \le \exp\left(-\frac{2n\varepsilon^2}{(b-a)^2}\right) .
$$

Compare the resulting sample complexity with Chebyshev's.

**Intuition.** Chebyshev spends a single moment of the sum; exponentiating first lets the whole moment generating function be used, and independence turns that into an $n$-th power.

**Solution**

**Step 1 — the Chernoff device.** For any $\lambda \gt 0$, monotonicity of $x \mapsto e^{\lambda x}$ and Markov's inequality give

$$
P\left(\sum_{i=1}^n (X_i - \mu) \ge n\varepsilon\right) = P\left(e^{\lambda\sum_i (X_i-\mu)} \ge e^{\lambda n \varepsilon}\right) \le e^{-\lambda n \varepsilon}\, E\!\left[e^{\lambda \sum_i (X_i-\mu)}\right] .
$$

By independence the expectation factorizes into $\prod_i E\left[e^{\lambda(X_i-\mu)}\right] = \left(E\left[e^{\lambda(X_1-\mu)}\right]\right)^n$. Raising to a power is where the exponential gain over Chebyshev originates: Chebyshev uses one moment of the sum, Chernoff uses the whole moment generating function and then optimizes.

**Step 2 — Hoeffding's lemma.** Claim: if $Y$ has mean $0$ and $Y \in [a - \mu, b - \mu]$ with range $R = b-a$, then

$$
E\left[e^{\lambda Y}\right] \le \exp\left(\frac{\lambda^2 R^2}{8}\right) .
$$

*Proof.* Convexity of $e^{\lambda y}$ on the interval gives, writing $y$ as a convex combination of the endpoints $\alpha = a-\mu$ and $\beta = b - \mu$,

$$
e^{\lambda y} \le \frac{\beta - y}{\beta - \alpha}e^{\lambda \alpha} + \frac{y - \alpha}{\beta-\alpha}e^{\lambda\beta} .
$$

Taking expectations and using $E[Y] = 0$ yields $E[e^{\lambda Y}] \le \frac{\beta}{\beta - \alpha}e^{\lambda\alpha} - \frac{\alpha}{\beta-\alpha}e^{\lambda\beta} =: e^{\psi(\lambda)}$. Setting $p = -\alpha/(\beta-\alpha) \in [0,1]$ and $u = \lambda(\beta-\alpha)$,

$$
\psi(\lambda) = -pu + \ln\left(1 - p + pe^{u}\right) ,
$$

which satisfies $\psi(0) = \psi'(0) = 0$ and $\psi''(u) = \frac{p e^u(1-p)}{(1-p+pe^u)^2} \le \frac{1}{4}$ (the maximum of $s(1-s)$ over $s \in [0,1]$). Taylor with remainder gives $\psi(\lambda) \le u^2/8 = \lambda^2R^2/8$. $\square$

**Step 3 — optimize $\lambda$.** Combining,

$$
P\left(\bar X_n - \mu \ge \varepsilon\right) \le \exp\left(-\lambda n \varepsilon + \frac{n\lambda^2 R^2}{8}\right) .
$$

The exponent is a quadratic in $\lambda$, minimized at $\lambda^\star = 4\varepsilon/R^2$, giving exponent $-2n\varepsilon^2/R^2$:

$$
P\left(\bar X_n - \mu \ge \varepsilon\right) \le \exp\left(-\frac{2n\varepsilon^2}{(b-a)^2}\right) . \qquad \blacksquare
$$

**Sample complexity.** To achieve failure probability $\delta$ (two-sided) Hoeffding needs $n \ge \frac{R^2\ln(2/\delta)}{2\varepsilon^2}$, versus Chebyshev's $n \ge \frac{\sigma^2}{\delta\varepsilon^2}$. The dependence on $\delta$ improves from $1/\delta$ to $\ln(1/\delta)$, which is what makes union bounds over $\vert \mathcal{H}\vert$ hypotheses affordable: the cost becomes $\ln\vert\mathcal{H}\vert$ rather than $\vert\mathcal{H}\vert$.

$$
\boxed{n_{\text{Hoeffding}} = O\!\left(\varepsilon^{-2}\log(1/\delta)\right) \text{ versus } n_{\text{Chebyshev}} = O\!\left(\varepsilon^{-2}\delta^{-1}\right)}
$$

*Key takeaway:* Exponentiating before applying Markov converts polynomial tail bounds into exponential ones — the foundation of generalization bounds and bandit algorithms.

### Problem L3.4 — When Does a Weighted Average Obey a CLT? (Lindeberg–Feller)

Let $Y_n = \sum_{i=1}^n w_{n,i}\,\xi_i$ where $\xi_i$ are i.i.d. with mean $0$, variance $1$, and $\sum_i w_{n,i}^2 = 1$. State the Lindeberg condition, show it reduces to $\max_i \vert w_{n,i}\vert \to 0$ for bounded $\xi$, and analyze the exponentially weighted average $w_{n,i} \propto \gamma^{n-i}$.

**Intuition.** The Lindeberg condition is a "no single term dominates" requirement, so for a weighted average it is really a statement about the largest weight in the array.

**Solution**

**Setup.** The triangular array $\{X_{n,i} = w_{n,i}\xi_i\}$ has row sums $Y_n$ with $E[Y_n] = 0$ and $\operatorname{Var}(Y_n) = \sum_i w_{n,i}^2 = 1$.

**Lindeberg condition.** For every $\varepsilon \gt 0$,

$$
L_n(\varepsilon) = \sum_{i=1}^n E\left[X_{n,i}^2\,\mathbf{1}\left\{\vert X_{n,i}\vert \gt \varepsilon\right\}\right] \longrightarrow 0 .
$$

**Theorem (Lindeberg–Feller).** If $L_n(\varepsilon) \to 0$ for all $\varepsilon \gt 0$, then $Y_n \xrightarrow{d} \mathcal{N}(0,1)$. The condition says no single term contributes a non-vanishing share of the variance — the sum must be a genuine aggregation of many comparable pieces.

**Reduction for bounded summands.** If $\vert \xi_i\vert \le M$, then $\vert X_{n,i}\vert \le M\vert w_{n,i}\vert \le M\,w_n^\star$ where $w_n^\star = \max_i \vert w_{n,i}\vert$. Whenever $M w_n^\star \lt \varepsilon$ the indicator is identically zero, so $L_n(\varepsilon) = 0$. Hence

$$
w_n^\star = \max_{i \le n} \vert w_{n,i}\vert \to 0 \quad \Longrightarrow \quad Y_n \xrightarrow{d}\mathcal{N}(0,1) .
$$

For the plain average $w_{n,i} = 1/\sqrt{n}$, $w_n^\star = n^{-1/2}\to 0$ — recovering the classical CLT.

**Exponentially weighted average.** Take $w_{n,i} = c_n\gamma^{\,n-i}$ with fixed $\gamma \in (0,1)$, normalized so that

$$
c_n^2\sum_{k=0}^{n-1}\gamma^{2k} = c_n^2\,\frac{1-\gamma^{2n}}{1-\gamma^2} = 1 \quad \Longrightarrow \quad c_n \to \sqrt{1-\gamma^2} .
$$

The largest weight is the most recent one, $w_{n,n} = c_n \to \sqrt{1-\gamma^2} \gt 0$, so $w_n^\star \not\to 0$ and the Lindeberg condition **fails**. Indeed the limit is $Y_\infty = \sqrt{1-\gamma^2}\sum_{k\ge 0}\gamma^k \xi_{n-k}$, an infinite weighted sum whose law is *not* Gaussian unless $\xi$ already is — it is a convolution dominated by the few most recent terms.

**Consequence for optimizers.** Momentum and Adam maintain exponentially weighted gradient averages with $\gamma \approx 0.9$–$0.999$. Their effective sample size is

$$
n_{\text{eff}} = \frac{1}{\sum_i w_{n,i}^2 / \left(\sum_i w_{n,i}\right)^2}\Bigg|_{\text{normalized}} \approx \frac{1+\gamma}{1-\gamma} ,
$$

roughly $19$ for $\gamma = 0.9$ and $1999$ for $\gamma = 0.999$. Noise reduction saturates at that finite level no matter how long training runs — momentum averages a *window*, not the whole history.

$$
\boxed{\text{CLT holds iff } \max_i \vert w_{n,i}\vert \to 0; \ \text{EMA has } n_{\text{eff}} \approx \tfrac{1+\gamma}{1-\gamma} \text{ and no CLT}}
$$

*Key takeaway:* Universality requires democratic weights; whenever a few terms dominate the variance the Gaussian limit disappears and the effective sample size stays bounded.

Compute the largest normalized weight and the effective sample size of an exponentially
weighted average.

In [14]:
n_w = 20_000
print(" gamma   max weight   sqrt(1 - gamma^2)   n_eff = (1+g)/(1-g)")
for g in [0.9, 0.99, 0.999]:
    w = g ** np.arange(n_w)[::-1]
    w = w / np.sqrt(np.sum(w**2))
    print(f"{g:6.3f}   {w.max():10.5f}   {np.sqrt(1 - g**2):17.5f}   "
          f"{(1 + g) / (1 - g):19.1f}")
    assert abs(w.max() - np.sqrt(1 - g**2)) < 1e-6

w_plain = np.full(n_w, 1 / np.sqrt(n_w))
print(f"\nplain average: max weight = {w_plain.max():.6f} -> 0, Lindeberg condition holds")
assert abs((1 + 0.9) / (1 - 0.9) - 19) < 1e-9

 gamma   max weight   sqrt(1 - gamma^2)   n_eff = (1+g)/(1-g)
 0.900      0.43589             0.43589                  19.0
 0.990      0.14107             0.14107                 199.0
 0.999      0.04471             0.04471                1999.0

plain average: max weight = 0.007071 -> 0, Lindeberg condition holds


The largest weight of the exponential scheme converges to $\sqrt{1-\gamma^2} \gt 0$ instead of to
zero, so the Lindeberg condition fails at every $\gamma \lt 1$ and no CLT applies. The effective
sample sizes $19$, $199$ and $1999$ are the noise floors that momentum and Adam cannot go below,
however long training runs; the plain average, whose maximum weight is $n^{-1/2} \to 0$, has no
such floor.